app/ts/components/pages/SettingsView.tsx

```diff
@@ -22,6 +26,74 @@ function CheckBoxSetting(param: CheckBoxSettingParam) {
 	)
 }
 
+function ImportExport() {
+	const [settingsReply, setSettingsReply] = useState<ImportSettingsReply | undefined>(undefined)
+
+	useEffect(() => {
+		async function popupMessageListener(msg: unknown) {
+			const message = ExternalPopupMessage.parse(msg)
+			if (message.method === 'popup_initiate_export_settings_reply') return setSettingsReply(message)
+			if (message.method !== 'popup_initiate_export_settings') return
+			downloadFile('interceptorSettingsAndAddressbook.json', message.data.fileContents)
+		}
+		browser.runtime.onMessage.addListener(popupMessageListener)
+
+		return () => browser.runtime.onMessage.removeListener(popupMessageListener)
+	})
+
+	const downloadFile = function (filename: string, fileContents: string) {
+		window.URL = window.webkitURL || window.URL
+		const blobData = new Blob([fileContents], { type: 'text/json; charset = utf-8' })
+		const a = document.createElement('a')
+		a.href = window.URL.createObjectURL(blobData)
+		a.download = filename
+		a.style.display = 'none'
+		document.body.appendChild(a)
+		a.click()
+		document.body.removeChild(a)
+	}
+
+	const importSettings = async (inputElement: { target: EventTarget | EventTarget & { files: FileList } | null }) => {
+		if (inputElement.target === null) return
+		if (!('files' in inputElement.target)) throw new Error('Did not select one file.')
+		if (inputElement.target.files.length !== 1) throw new Error('Did not select one file.')
+		const reader = new FileReader()
+		reader.readAsText(inputElement.target.files[0])
+		reader.onloadend = async function() {
+			if (reader.result === null) throw new Error('failed to load file')
+			await sendPopupMessageToBackgroundPage({ method: 'popup_import_settings', data: { fileContents: reader.result as string } })
+		}
+		reader.onerror = function() {
+			console.error(reader.error)
+			throw new Error('error on importing settings')
+		}
+	}
+	const exportSettings = async () => await sendPopupMessageToBackgroundPage({ method: 'popup_get_export_settings' })
+	return <>
+		{ settingsReply !== undefined && settingsReply.data.success === false ?
+			<div style = 'margin: 10px; background-color: var(--bg-color);'>
+				<ErrorComponent warning = { true } text = { settingsReply.data.errorMessage }/>
+			</div>
+		: <></> }
+		{ settingsReply !== undefined && settingsReply.data.success === true ?
+			<div style = 'margin: 10px; background-color: var(--bg-color);'>
+				<DinoSays text = { 'Settings and address book loaded!' }/>
+			</div>
+		: <></> }
+		<div class = 'popup-button-row'>
+			<div style = 'display: flex; flex-direction: row;'>
+				<button className = 'button is-primary is-danger' style = 'flex-grow: 1; margin-left: 5px; margin-right: 5px;'>
+					Import settings
+					<input type = 'file' accept = '.json' onInput = { importSettings } style = 'position: absolute; width: 100%; height: 100%;'/>
+				</button>
```

Comment:
I don't think this should be a button, as it doesn't function like a button.  The input is what matters here.  I also think you should just put the opacity here rather than in CSS since this is basically the only place it applies.  CSS is useful (for readability and performance) when you need to apply the same style to many elements.  However, if the style only applies to one element then inline helps with readability IMO.
```suggestion
				<label className = 'button is-primary is-danger' style = 'flex-grow: 1; margin-left: 5px; margin-right: 5px;'>
					Import settings
					<input type = 'file' accept = '.json' onInput = { importSettings } style = 'position: absolute; width: 100%; height: 100%; opacity: 0;'/>
				</label>
```

Reply:
hmm. I feel it functions like a button. You press it and file selection dialog appears

Reply:
But you aren't clicking the button, you are clicking the `Input` field.  The button doesn't actually behave like a button, it behaves like a label in the current code.  This is, in theory, particularly important for screen readers and such which will read this as a button wrapping an input field which is confusing.  A labelled input field on the other hand is much more clear as to what the association is.

Reply:
ah, yeah. I guess a better way would be to make the input look like a button instead with CSS

Reply:
Yes, that would be ideal.  The problem is that styling `input` fields is *really* hard.  You can do it, but it is a huge pain.

Reply:
and you want me to go through huge pain, instead of using tricks like this? D:

Reply:
😆 no.  I actually started writing a comment suggesting that you style the input as a button but realized it would be too hard which is why I switched to just suggesting swapping the `button` element into a `label` element.